lib/puppetserver/ca/action/delete.rb

```diff
@@ -0,0 +1,187 @@
+          errored = false
+          deleted_count = 0
+          inventory_file_path = File.join(settings[:cadir], 'inventory.txt')
+
+          if args['expired']
+            # Delete expired certs found in inventory first since this is cheaper.
+            # Then, look for any certs not in the inventory, check if they
+            # are expired, then delete those.
+            inventory, err = Inventory.parse_inventory_file(inventory_file_path, @logger)
+            errored ||= err
+            expired_in_inventory = inventory.select { |k,v| v[:not_after] < Time.now }.map(&:first)
+            count, err = delete_certs(settings[:cadir], expired_in_inventory)
+            deleted_count += count
+            errored ||= err
+            other_certs_to_check = find_certs_not_in_inventory(settings[:cadir], inventory.map(&:first))
+            count, err = delete_expired_certs(settings[:cadir], other_certs_to_check)
+            deleted_count += count
+            errored ||= err
+          end
+
+          plural = deleted_count == 1 ? "" : "s"
+          @logger.inform("#{deleted_count} certificate#{plural} deleted.")
+          # If encountered non-fatal errors (an invalid entry in inventory.txt, cert not existing on disk)
+          # return 24. Returning 1 should be for fatal errors where we could not do any part of the action.
+          return errored ? 24 : 0
+        end
+
+        def find_certs_not_in_inventory(cadir, inventory_certnames)
+          all_cert_files = Dir.glob("#{cadir}/signed/*.pem").map { |f| File.basename(f, '.pem') }
+          all_cert_files - inventory_certnames
+        end
+
+        def delete_certs(cadir, certnames)
+          deleted = 0
+          errored = false
+          certnames.each do |cert|
+            path = "#{cadir}/signed/#{cert}.pem"
+            if File.exist?(path)
+              @logger.inform("Deleting certificate at #{path}")
+              File.delete(path)
+              deleted += 1
+            else
+              @logger.err("Could not find certificate file at #{path}")
+              errored = true
+            end
+          end
+          [deleted, errored]
+        end
+
+        def delete_expired_certs(cadir, certnames)
+          deleted = 0
+          errored = false
+          files = certnames.map { |c| "#{cadir}/signed/#{c}.pem" }
+          files.each do |f|
+            # Shouldn't really be possible since we look for certs on disk
+            # before calling this function, but just in case.
+            unless File.exist?(f)
+              @logger.err("Could not find certificate file at #{f}")
+              errored = true
+              next
+            end
+            begin
+              cert = OpenSSL::X509::Certificate.new(File.read(f))
+            rescue OpenSSL::X509::CertificateError
+              @logger.err("Error reading certificate at #{f}")
+              errored = true
+              next
+            end
+            if cert.not_after < Time.now
```

Comment:
Should this be converted to utc?

Reply:
I believe it's converted to a plain `Time` object so it can be directly compared.  I'll double check, though.

Reply:
Yeah, confirmed, it evaluates it correctly.  In the case below, I created a localized time at the +8 timezone that on its face looks later than `cert.not_after`, but when converted to UTC, is before `cert.not_after`.  It is evaluated correctly.

```
[54] pry(#<Puppetserver::Ca::Action::Delete>)> cert.not_after
=> 2038-02-20 19:11:48 UTC
[55] pry(#<Puppetserver::Ca::Action::Delete>)> t=Time.parse('2038-02-21 00:00:00.0000000 +0800')
=> 2038-02-21 00:00:00 +0800
[56] pry(#<Puppetserver::Ca::Action::Delete>)> t.utc
=> 2038-02-20 16:00:00 UTC
[57] pry(#<Puppetserver::Ca::Action::Delete>)> cert.not_after < t
=> false```